### PR TITLE
zitadel: init at 2.37.2; nixos/zitadel: init

### PR DESCRIPTION
**Pull request by Sorixelle**
## Description of changes

[ZITADEL](https://zitadel.com/) is a identity and access management platform. This PR creates a Nix derivation for it, as well as a NixOS module for running it as a service.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Sorixelle** [comment]
Result of `nixpkgs-review pr 254896` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zitadel</li>
  </ul>
</details>

File: nixos/modules/services/web-apps/zitadel.nix [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+let
+  cfg = config.services.zitadel;
+
+  settingsFormat = pkgs.formats.yaml { };
+in {
+  options.services.zitadel =
+    let inherit (lib) mdDoc mkEnableOption mkOption mkPackageOptionMD types;
+    in {
+      enable = mkEnableOption
+        (mdDoc "ZITADEL, a user and identity access management platform.");
```
Comment (Gabriella439) [inline nixos/modules/services/web-apps/zitadel.nix:12]: I believe the `mdDoc` is no longer necessary.  Based on my cursory understanding it was only added as a transitional function and now it's just a synonym for `id`.

```suggestion
    let inherit (lib) mkEnableOption mkOption mkPackageOptionMD types;
    in {
      enable = mkEnableOption
        "ZITADEL, a user and identity access management platform.";
```

Same comment for all of the other `md` utilities (like `mkPackageOptionMD`, which you can now replace with `mkPackageOption`)

File: nixos/modules/services/web-apps/zitadel.nix [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        '';
+      };
+    };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.tlsMode != "enabled"
+        || ((cfg.settings.TLS.Key != null || cfg.settings.TLS.KeyPath != null)
+          && (cfg.settings.TLS.Cert != null || cfg.settings.TLS.CertPath
+            != null));
```
Comment (Gabriella439) [inline nixos/modules/services/web-apps/zitadel.nix:176]: Minor suggestion:

```suggestion
      assertion = cfg.tlsMode == "enabled"
        -> ((cfg.settings.TLS.Key != null || cfg.settings.TLS.KeyPath != null)
          && (cfg.settings.TLS.Cert != null || cfg.settings.TLS.CertPath
            != null));
```

File: nixos/modules/services/web-apps/zitadel.nix [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      path = [ cfg.package ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = ''
+        zitadel start-from-init \
+          ${configArgs} \
+          ${stepsArgs} \
+          --masterkeyFile ${cfg.masterKeyFile} \
+          --tlsMode ${cfg.tlsMode}
+      '';
```
Comment (Gabriella439) [inline nixos/modules/services/web-apps/zitadel.nix:205]: I think you can simplify some of this using `lib.toGNUCommandLine`, like this:

```suggestion
    in {
      description = "ZITADEL identity access management";
      path = [ cfg.package ];
      wantedBy = [ "multi-user.target" ];

      script = ''
        zitadel start-from-init ${lib.toGNUCommandLine { } {
          config = cfg.extraSettingsPaths ++ [ configFile ];
          steps = cfg.extraSettingsPaths ++ [ stepsFile ];
          masterkeyFile = cfg.masterKeyFile;
          inherit (cfg) tlsMode;
      }'';
```

**Review by Gabriella439 — COMMENTED** [review]
Looks great!  Just a few small suggestions

**Comment by pinpox** [comment]
Anything missing on this? I'm eager to try it out :+1:

File: nixos/modules/services/web-apps/zitadel.nix [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      let
+        configFile = settingsFormat.generate "config.yaml" cfg.settings;
+        stepsFile = settingsFormat.generate "steps.yaml" cfg.steps;
+
+        args = builtins.concatStringsSep " " (lib.cli.toGNUCommandLine { } {
+          config = cfg.extraSettingsPaths ++ [ configFile ];
+          steps = cfg.extraStepsPaths ++ [ stepsFile ];
+          masterkeyFile = cfg.masterKeyFile;
+          inherit (cfg) tlsMode;
+        });
```
Comment (Gabriella439) [inline nixos/modules/services/web-apps/zitadel.nix:196]:
```suggestion
        args = lib.cli.toGNUCommandLineShell { } {
          config = cfg.extraSettingsPaths ++ [ configFile ];
          steps = cfg.extraStepsPaths ++ [ stepsFile ];
          masterkeyFile = cfg.masterKeyFile;
          inherit (cfg) tlsMode;
        };
```

**Review by Gabriella439 — COMMENTED** [review]
Just one last thing.  After you fix this and resolve the merge conflict I'm ready to merge this

File: pkgs/by-name/zi/zitadel/package.nix [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    go generate internal/api/ui/login/static/resources/generate.go
+    go generate internal/api/ui/login/statik/generate.go
+    go generate internal/notification/statik/generate.go
+    go generate internal/statik/generate.go
+
+    mkdir -p docs/apis/assets
+    go run internal/api/assets/generator/asset_generator.go -directory=internal/api/assets/generator/ -assets=docs/apis/assets/assets.md
+
+    cp -r ${passthru.console}/* internal/api/ui/console/static
+    CGO_ENABLED=0 go build -o zitadel -v -ldflags="-s -w -X 'github.com/zitadel/zitadel/cmd/build.version=${version}'"
```
Comment (Aleksanaa) [inline pkgs/by-name/zi/zitadel/package.nix:125]: Is there a good reason not to use [`ldflags`](https://ryantm.github.io/nixpkgs/languages-frameworks/go/#var-go-ldflags)?

Now if I want to make a debug build of zitadel, I have to use `nix build --impure --expr '(import <nixpkgs> {}).zitadel.overrideAttrs (old: { buildPhase = builtins.replaceStrings [ "-s -w" ] [ "" ] old.buildPhase; })'`.

**Pull request merged**